doc/BGP/BGP-supress-fib-pending.md

```diff
@@ -0,0 +1,804 @@
+  4. Verify RTM_F_OFFLOAD set in route message sent  to zebra
+- FPMSyncd: send RTM_DELROUTE
+  1. Mock FPM interface
+  2. Call RouteSync::onRouteResponse() with delete response message containing route 1.1.1.0/24
+  4. Verify no message is sent to zebra
+
+- RouteOrch: program ASIC route
+  1. Create FVS and call RouteOrch::doTask()
+  2. Verify route is created in SAI
+  3. Verify the content of APPL_STATE_DB
+- RouteOrch: delete ASIC route
+  1. Create delete entry and call RouteOrch::doTask()
+  2. Verify route is delete in SAI
+  3. Verify the content of APPL_STATE_DB
+
+#### 11.2. VS Test cases
+
+Code coverage is satisfied by UT and the E2E flow is not possible to test with current VS infrastructure, so no VS tests are planned.
+
+#### 11.3. System Test cases
+
+<!-- omit in toc -->
+##### Figure 7. T1 test topology
+
+<p align=center>
+<img src="img/test-topology.svg" alt="Figure 7. T1 test topology">
+</p>
+
+In order to test this feature end to end it is required to simulate a delay in ASIC route programming. The proposed test:
+
+- **Functional Test**
+  1. Enable ```suppress-fib-pending```
+  2. Suspend orchagent process to simulate a delay with ```kill --SIGSTOP $(pidof orchagent)```
+  3. Announces prefix *1.1.1.0/24* (*1::/64*) to DUT through *exabgp* from T0 Arista VM
+  4. Make sure announced BGP route is queued in ```show ip route``` output
+  5. Verify the route is not announced to Arista T2 peer by executing ```show ip bgp neighbor A.B.C.D received-routes``` on the peer
+  6. Send traffic matching the prefix and verify packets are not forwarded to T0 Arista VM
+  7. Restore orchagent process: ```kill --SIGCONT $(pidof orchagent)```
+  8. Make sure route is programmed in ASIC_DB
+  9. Make sure announced BGP route is FIB synced in ```show ip route``` output
+  10. Verify the route is announced to Arista T1 peer by executing ```show ip bgp neighbor A.B.C.D received-routes``` on the peer
+  11. Send traffic matching the prefix and verify packets are forwarded to T0 Arista VM
+- **Performance Test**
+  1. Announce 10K routes to DUT
+  2. Send traffic towards these prefixes from AUX
+  3. Measure time from announce start till all packets get received
+  4. The test is parametrized for suppression state for comparison
+  5. The test passes if the time is under the predefined threshold
+- **Stress Test**
+  1. Enable ```suppress-fib-pending```
+  2. In a loop for 100 times:
+     1. Announce 10 prefix to DUT through *exabgp* from T0 Arista VM
+     2. Withdraw 10 prefix to DUT through *exabgp* from T0 Arista VM
+  3. Once reached the required number of cycles the loop breaks after first step
+  4. Consistency check is applied, there are no withdrawn routes and announced routes were successfully installed and correctly marked as offloaded in zebra (In case a race condition happens or a notification is missed for some reason this test will try to catch it)
+
```

Comment:
Don't we have any test case to simulate orchagent crash and delay the response to fpmsyncd from route-orch?

Reply:
In case orchagent crash there will be no response from routeorch. Which tests case do you propose?

Reply:
What will happen if fpmsyncd doesnot notify RTM_F_OFFLOAD to Zebra due to orchagent crash? Is there any timeout in Zebra to declare FIB programming failed? what would happen for those routes in Zebra that are waiting for the FIB programming after OrchAgent back up?

Reply:
There's no logic in zebra that considers a route beeing failed in HW if it did not receive RTM_F_OFFLOAD flag within some time interval. These routes will stay in _queue_-ed state.
Orchagent crash itself triggers BGP container(zebra + bgpd) restart. So in that case, zebra starts fresh with no routes.